node/syncblock.go

```diff
@@ -368,7 +406,44 @@ func (localNode *LocalNode) initSyncing() {
 // StartSyncing starts block syncing from current local ledger until it gets to
 // block height stopHeight with block hash stopHash from given neighbors
 func (localNode *LocalNode) StartSyncing(syncStopHash common.Uint256, syncStopHeight uint32, neighbors []*RemoteNode) (bool, error) {
-	var err error
+	cs := chain.DefaultLedger.Store.(*store.ChainStore)
+
+	fastSyncHeight, err := cs.GetSyncRootHeight()
+	if err != nil {
+		log.Info("local sync root height not found:", err)
+	}
+	if config.Parameters.SyncMode != "fast" && fastSyncHeight > cs.GetHeight() {
+		log.Fatalf("invalid sync mode %s at height %d", config.Parameters.SyncMode, cs.GetHeight())
+	}
+
+	if config.Parameters.SyncMode == "fast" && cs.ShouldFastSync() {
+		if fastSyncHeight == 0 {
+			fastSyncHeight = syncStopHeight - config.Parameters.RecentStateCount
+		}
+		fastSyncRootHash, err := cs.GetFastSyncStateRoot()
+		if err != nil {
+			fastSyncRootHash = localNode.GetNeighborsMajorityStateRootByHeight(fastSyncHeight, neighbors)
+			if fastSyncRootHash == common.EmptyUint256 {
+				log.Fatalf("get neighbors majority state root of height %d failed", fastSyncHeight)
```

Comment:
No need to fatal, when put inside syncOnce.Do, just set err and return, the caller will falta